### ISSUE TITLE
Support for IOT Rule Action "dynamoDBv2"

### ISSUE THREAD

**Issue by tobiasschaber** [terraform-providers/terraform-provider-aws#4534]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

There is already a dynamodb IOT topic rule ("aws_iot_topic_rule"), but the dynamodb version 2 rule is not yet suported. It would be nice to have that also.

The version 2.0 rule action is called something like "Split message into multiple columns of a database table (DynamoDB Version 2)" and it's available via AWS Management console.

### New or Affected Resource(s)

* aws_iot_topic_rule, dynamodb object

### Potential Terraform Configuration

```
table_name: my_target_table
iam_role: my_iam_role
```
### References
* https://docs.aws.amazon.com/iot/latest/developerguide/iot-rule-actions.html
See "dynamoDBv2" rule

**Comment by flmag** [comment]
Hi,

Also need dynamodbv2 for some projects, any updates ?

Regards,
Florian

**Comment by gmarkey** [comment]
I've hacked together support for this as I needed it in the meantime. I do not recommend using this patch if you want your statefile to be valid once support for DDBv2 rules is properly added.

https://gist.github.com/gmarkey/1fe21a6d16b35d8b73c2600b73684f74

**Comment by abdelhakim-atmani** [comment]
Hi,

Any update ?

**Comment by legokichi** [comment]
here is my workaround

```tf
locals {
    prefix = "test-iot-action-dynamodb2"
    topic = "${local.prefix}/statuslog"
    rule = "${replace("${local.prefix}-rule", "-", "_")}"
}

resource "aws_dynamodb_table" "dynamodb_table" {
    name           = "${local.prefix}-table"

    read_capacity  = 2
    write_capacity = 2
    hash_key       = "DeviceId"
    range_key      = "TimeStamp"

    attribute = [{
        name = "DeviceId"
        type = "S"
    }, {
        name = "TimeStamp"
        type = "N"
    }]
}

resource "aws_iam_role" "iot_rule" {
    name = "${local.prefix}-iot_rule"
    assume_role_policy = <<EOF
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Allow",
            "Principal": {
                "Service": "iot.amazonaws.com"
            },
            "Action": "sts:AssumeRole"
        }
    ]
}
EOF
}

data "template_file" "iot_rule" {
    template = <<EOF
{
    "Version": "2012-10-17",
    "Statement": {
        "Effect": "Allow",
        "Action": "dynamodb:PutItem",
        "Resource": "$${table_arn}"
    }
}
EOF
    vars {
        table_arn = "${aws_dynamodb_table.dynamodb_table.arn}"
    }
}

resource "aws_iam_policy" "iot_rule" {
    name = "${local.prefix}-iot_rule"
    policy = "${data.template_file.iot_rule.rendered}"
}

resource "aws_iam_role_policy_attachment" "iot_rule" {
    role = "${aws_iam_role.iot_rule.name}"
    policy_arn = "${aws_iam_policy.iot_rule.arn}"
}

# resource "aws_iot_topic_rule" "iot_rule" {
#     # MUST be ^[a-zA-Z0-9_]+$
#     name = "${replace("${local.rule}", "-", "_")}"
#     sql = "SELECT * FROM '${local.topic}'"
#     sql_version = "2016-03-23"

#     dynamodbv2 {
#         role_arn = "${aws_iam_role.iot_rule.arn}"
#         table_name = "${aws_dynamodb_table.dynamodb_table.arn}"
#     }
# }


data "template_file" "iot_rule_patch" {
    template = <<EOF
{
    "sql": "SELECT * FROM '${local.topic}'",
    "awsIotSqlVersion": "2016-03-23",
    "actions": [
        {
            "dynamoDBv2": {
                "roleArn": "$${role_arn}",
                "putItem": {
                    "tableName": "$${table_name}"
                }
            }
        }
    ]
}
EOF
    vars {
        role_arn = "${aws_iam_role.iot_rule.arn}"
        table_name = "${aws_dynamodb_table.dynamodb_table.arn}"
    }
}

resource "null_resource" "iot_rule_patch" {
    provisioner "local-exec" {
        command = "(aws iot delete-topic-rule --rule-name ${local.rule} || : ) && aws iot create-topic-rule --rule-name ${local.rule} --topic-rule-payload '${data.template_file.iot_rule_patch.rendered}'"
    }
    depends_on = [
        "aws_iam_role_policy_attachment.iot_rule"
    ]
}
```

**Comment by Martinnew** [comment]
Thanks for your workaround! 
To make it work I had to give the dynamodb_table name, instead the arn to the template_file. 

Now I ran into another problem: I simply cannot use the  '+' wildcard character within the sql statement. As I make heavy use of wildcards for our setup this makes this attempt pretty much useless for me.

**Comment by bflad** [comment]
Support for this functionality has been merged and will release with version 2.62.0 of the Terraform AWS Provider, later this week. Thanks to @edwardbrowncross for the implementation. 👍
